src/west/cmd/project.py

```diff
@@ -480,56 +480,67 @@ def _manifest_path(args):
+    # project's repository does not already exist, it is created first.
     #
-    # Returns True if the project already existed, and False if it was cloned.
+    # Returns True if the project's repository already existed.
 
-    if _cloned(project):
-        existed = True
-        _inf(project, 'Fetching changes for (name-and-path)')
-        _git(project, 'remote update')
-    else:
-        existed = False
+    exists = _cloned(project)
+
+    if not exists:
+        _inf(project, 'Creating repository for (name-and-path)')
+        _git_base(project, 'init (abspath)')
+        _git(project, 'remote add origin (url)')
 
-        # --depth implies --single-branch, so we must pass --branch when
-        # creating a shallow clone (since the remote might not have HEAD
-        # pointing to the 'revision' branch).
+    if project.clone_depth:
+        _inf(project,
+             'Fetching changes for (name-and-path) with --depth (clone-depth)')
+
+        # If 'clone-depth' is specified, fetch just the specified revision
+        # (probably a branch). That will download the minimum amount of data,
+        # which is probably what's wanted whenever 'clone-depth is used. The
+        # default 'git fetch' behavior is to do a shallow clone of all branches
+        # on the remote.
         #
-        # Currently, we also pass --branch for non-shallow clones, which checks
-        # the 'revision' branch out in the newly created repo, but this might
-        # change.
-
-        msg = 'Cloning (name-and-path)'
-        cmd = 'clone'
-        if _is_sha(project.revision):
-            # If the project revision is a SHA and the remote repo's
-            # HEAD doesn't point to anything valid, the clone
-            # operation might fail to check anything out. In that
-            # case, make sure to check out the given revision.
-            checkout = True
-        else:
-            cmd += ' --branch (revision)'
-            checkout = False
-        if project.clone_depth:
-            msg += ' with --depth (clone-depth)'
-            cmd += ' --depth (clone-depth)'
-        cmd += ' (url) (path)'
+        # Note: Many servers won't allow fetching arbitrary commits by SHA.
+        # Combining --depth with an SHA will break for those.
 
-        _inf(project, msg)
-        _git_base(project, cmd)
+        # Qualify branch names with refs/heads/, just to be safe. Just the
+        # branch name is likely to work as well though.
+        _git(project,
+             'fetch --depth=(clone-depth) origin ' +
+                 (project.revision if _is_sha(project.revision) else \
+                     'refs/heads/' + project.revision))
 
-        if checkout:
-            _checkout(project, project.revision)
+    else:
+        _inf(project, 'Fetching changes for (name-and-path)')
+
+        # If 'clone-depth' is not specified, fetch all branches on the
+        # remote. This gives a more usable repository.
```

Comment:
Shouldn't break as long as the SHA is reachable from any branch on the remote, which is probably reasonable assumption.

The comment a bit above this code is relevant here too:

```
# Note: Many servers won't allow fetching arbitrary commits by SHA.
# ...
```

Reply:
> Shouldn't break as long as the SHA is reachable from a branch on the remote, which is probably a reasonable assumption.

I agree this is an edge case, but why not handle it, since it's simple to do so?

> The comment a bit above this code is relevant here too:
> 
> # Note: Many servers won't allow fetching arbitrary commits by SHA.
> # ...

Sounds like a server bug. The git fetch manpage on my system allows it:

```
           The format of a <refspec> parameter is an optional plus +, followed
           by the source <src>, followed by a colon :, followed by the
           destination ref <dst>. The colon can be omitted when <dst> is
           empty. <src> is typically a ref, but it can also be a fully spelled
           hex object name.
```

If that happens, the user can work around their buggy Git server by providing a branch or tag ref, no?

Reply:
> > Shouldn't break as long as the SHA is reachable from a branch on the remote, which is probably a reasonable assumption.
> 
> I agree this is an edge case, but why not handle it, since it's simple to do so?

It's also by design. If you aren't specifying a clone depth, it probably means you're going to actively work on the repository, and a repository with all the branches/tags from the remote available is more useful. SHA revisions aren't special-cased.

> If that happens, the user can work around their buggy Git server by providing a branch or tag ref, no?

I think it's disabled on most servers, for security reasons perhaps. Looks like GitHub disables it on all repositories, though I haven't dug up a better source: https://gist.github.com/brunoguerra/3b1a1fb3d4e076bd8fa1d00046ceaed3.

I get that error for the main Zephyr repo and for Kconfiglib:

    $ git init
    $ git remote add origin https://github.com/ulfalizer/Kconfiglib.git
    $ git fetch origin 78682a8e3c4fefffec281f13168cec5c297827bb
    error: Server does not allow request for unadvertised object 78682a8e3c4fefffec281f13168cec5c297827bb

Reply:
Same thing happens when fetching from a repository on the local filesystem, so seems like a Git default in general.

Reply:
Think I found some relevant stuff by digging through Git sources (with a "TODO: Document this further", because IT...)

https://github.com/git/git/blob/2f743933341f276111103550fbf383a34dfcfd38/Documentation/technical/http-protocol.txt#L331

Looks like almost no one enables those.

Reply:
> Looks like almost no one enables those.

I guess our corp's git server is a rare exception. OK, I'll fix this in my repository and drop this complaint for west.